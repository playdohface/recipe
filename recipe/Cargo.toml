[package]
name = "recipe"
version = "0.1.0"
edition = "2021"

[dependencies]
convert_case = "0.6.0"
nom = "7.1.3"
nom_locate = "4.2.0"
serde_json = "1.0.132"
json5 = "0.4.1"
anyhow = "1.0.93"
handlebars = "6.2.0"
thiserror = "2.0.9"
itertools = "0.14.0"
serde = { version = "1.0.217", features = ["derive"] }

[dev-dependencies]
insta = { version = "1.41.1", features = ["yaml"] }
tempfile = "3.14.0"
