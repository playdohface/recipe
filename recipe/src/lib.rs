
pub mod context;
pub mod loader;
pub mod parser;
