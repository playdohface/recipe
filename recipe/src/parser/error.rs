use super::{Span, Token};

#[derive(Debug)]
pub enum ParseError<'a> {
    Generic(Span<'a>),
    InvalidToken(Token<'a>),
    InvalidCodeBlock(Span<'a>),
    NomError(nom::error::VerboseError<Span<'a>>),
}
