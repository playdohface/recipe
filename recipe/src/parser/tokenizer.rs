use std::ops::RangeFrom;

use anyhow::anyhow;
use convert_case::{Case, Casing};
use nom::{
    branch::alt,
    bytes::complete::{is_not, tag, take_till, take_while},
    character::complete::{char, line_ending, space0, space1},
    combinator::fail,
    multi::many1_count,
    sequence::{delimited, preceded, terminated},
    IResult, InputLength, Parser, Slice,
};
use nom_locate::position;

use crate::loader;

use super::{Heading, Keyword, Link, Span, Token, TokenType};

#[derive(Debug)]
pub struct Tokenizer<'a> {
    /// The path being tokenized
    pub origin_path: &'a str,
    src: Span<'a>,
    scope: usize,
}

impl<'a> Tokenizer<'a> {
    pub fn from_str(src: &'a str, path: &'a str) -> Self {
        Tokenizer {
            origin_path: path,
            src: src.into(),
            scope: 0,
        }
    }
    /// Only return the Tokens found inside a given heading
    /// The heading-slug is the kebab-case version of the heading text
    fn scope_to_heading(&mut self, heading_slug: &str) {
        loop {
            match self.next() {
                Some(Token {
                    inner: TokenType::Heading(Heading { text, level, .. }),
                    ..
                }) if text.to_case(Case::Kebab) == heading_slug => {
                    self.scope = level;
                    break;
                }
                Some(_) => continue,
                None => break,
            }
        }
    }
}
impl<'a> Iterator for Tokenizer<'a> {
    type Item = Token<'a>;

    fn next(&mut self) -> Option<Self::Item> {
        if let Ok((rest, found)) = find_next(any_token)(self.src) {
            match found.inner {
                TokenType::Heading(Heading { level, .. }) if level <= self.scope => {
                    return None;
                }
                _ => {
                    self.src = rest;
                    Some(found)
                }
            }
        } else {
            None
        }
    }
}

/// Eat input until the parser succeeds
fn find_next<I, O, F>(mut f: F) -> impl FnMut(I) -> IResult<I, O>
where
    F: FnMut(I) -> IResult<I, O>,
    I: Clone + InputLength + Slice<RangeFrom<usize>>,
{
    move |mut inp: I| {
        while inp.input_len() > 0 {
            if let Ok(found) = f.parse(inp.clone()) {
                return Ok(found);
            } else {
                inp = inp.slice(1..);
            }
        }
        fail(inp)
    }
}

fn any_token(inp: Span) -> IResult<Span, Token> {
    alt((
        parse_heading,
        parse_block,
        parse_inline,
        parse_link,
        parse_keyword,
        parse_newline,
    ))(inp)
}

// Parse a markdown heading
fn parse_heading(inp: Span) -> IResult<Span, Token> {
    let (inp, start) = position(inp)?;
    let (inp, level) = preceded(terminated(line_ending, space0), many1_count(char('#')))(inp)?;
    let (inp, _) = take_till(|c: char| !c.is_whitespace())(inp)?;
    let (inp, text) = is_not("\r\n")(inp)?;
    let (inp, _) = line_ending(inp)?;

    let heading = Heading {
        level,
        text: text.trim_end(),
    };
    Ok((
        inp,
        Token {
            start,
            end: position(inp)?.1,
            inner: TokenType::Heading(heading),
        },
    ))
}

/// Parse a raw codeblock (delimited by trtiple backticks)
fn parse_block(inp: Span) -> IResult<Span, Token> {
    let (rest, res) = delimited(tag("```"), take_while(|c| c != '`'), tag("```"))(inp)?;
    Ok((
        rest,
        Token {
            start: position(res)?.1,
            end: position(rest)?.1,
            inner: TokenType::Block(res.fragment()),
        },
    ))
}

fn parse_inline(inp: Span) -> IResult<Span, Token> {
    let (rest, res) = delimited(
        tag("`"),
        take_while(|c| c != '`' && !is_line_end(c)),
        tag("`"),
    )(inp)?;
    Ok((
        rest,
        Token {
            start: position(res)?.1,
            end: position(rest)?.1,
            inner: TokenType::Inline(res.fragment()),
        },
    ))
}

/// Parse a Markdown-link
fn parse_link(inp: Span) -> IResult<Span, Token> {
    let (inp, start) = position(inp)?;
    let (inp, text) = delimited(tag("["), take_while(is_allowed_in_link_text), tag("]"))(inp)?;
    let (inp, path) = delimited(tag("("), take_while(is_allowed_in_link_path), tag(")"))(inp)?;
    let link = TokenType::Link(Link {
        text: &text,
        path: &path,
    });
    Ok((
        inp,
        Token {
            start,
            end: position(inp)?.1,
            inner: link,
        },
    ))
}

fn is_allowed_in_link_text(c: char) -> bool {
    c != ']' && !is_line_end(c)
}

fn is_allowed_in_link_path(c: char) -> bool {
    c != ')' && !is_line_end(c)
}

fn is_line_end(c: char) -> bool {
    c == '\n' || c == '\r'
}

fn parse_keyword(inp: Span) -> IResult<Span, Token> {
    let (rest, res) = terminated(
        alt((
            tag("To"),      //
            tag("Set"),     //
            tag("Execute"), //
        )),
        alt((space1, line_ending)),
    )(inp)?;
    let keyword = match *res.fragment() {
        "To" => Keyword::To,
        "Set" => Keyword::Set,
        "Execute" => Keyword::Execute,
        _ => unreachable!(),
    };
    Ok((
        rest,
        Token {
            start: position(res)?.1,
            end: position(rest)?.1,
            inner: TokenType::Keyword(keyword),
        },
    ))
}

fn parse_newline(inp: Span) -> IResult<Span, Token> {
    let (rest, _newline) = line_ending(inp)?;
    Ok((
        rest,
        Token {
            start: position(inp)?.1,
            end: position(rest)?.1,
            inner: TokenType::Newline,
        },
    ))
}

#[cfg(test)]
mod tests {
    use nom::Offset;
    use nom_locate::LocatedSpan;

    use super::*;

    #[test]
    fn test_scope_to_heading() {
        let src = r#"
# Heading 1
`inside heading 1`
## Heading 2 please ScopeMe-to_ThisHeading
`inside heading 2`
### Heading 3
`inside heading 3`
## Another heading
`inside another heading`
"#;
        let mut tokenizer = Tokenizer::from_str(src, "Mock.md");
        tokenizer.scope_to_heading("heading-2-please-scope-me-to-this-heading");
        let tokens: Vec<TokenType> = tokenizer.map(|t| t.inner).collect();
        assert_eq!(
            tokens,
            vec![
                TokenType::Inline("inside heading 2"),
                TokenType::Heading(Heading {
                    level: 3,
                    text: "Heading 3"
                }),
                TokenType::Inline("inside heading 3"),
            ]
        );
    }

    #[test]
    fn test_tokenizer() {
        let src = r#"
#Heading 1
Some text
```rust
### Not a heading
fn main() {
    println!("Hello, World!");
}
```
Some more text
[Link](https://example.com)
Set `foo` to `bar`
"#;
        let tokenizer = Tokenizer::from_str(src, "Mock.md");
        let tokens: Vec<TokenType> = tokenizer.map(|t| t.inner).collect();
        assert_eq!(
            tokens,
            vec![
                TokenType::Heading(Heading {
                    level: 1,
                    text: "Heading 1"
                }),
                TokenType::Newline,
                TokenType::Block(
                    "rust\n### Not a heading\nfn main() {\n    println!(\"Hello, World!\");\n}\n"
                ),
                TokenType::Newline,
                TokenType::Newline,
                TokenType::Link(Link {
                    text: "Link",
                    path: "https://example.com"
                }),
                TokenType::Newline,
                TokenType::Keyword(Keyword::Set),
                TokenType::Inline("foo".into()),
                TokenType::Inline("bar".into()),
                TokenType::Newline,
            ]
        );
    }

    #[test]
    fn test_find_next() {
        let src = "foovrfs barand more";
        let (rest, res) = find_next(tag("bar"))(src).unwrap();
        assert_eq!(res, "bar");
        assert_eq!(rest, "and more");
        assert!(find_next(tag("foo"))("haystack with no match").is_err());
    }

    #[test]
    fn test_parse_heading() {
        let undertest = "\n #### \t How To Bake a Cake\t  \r\nand live to tell the tale";
        let expected = TokenType::Heading(Heading {
            level: 4,
            text: "How To Bake a Cake",
        });
        let (rest, parsed) = assert_parse_token_with_inner(&undertest, parse_heading, expected);
        assert_eq!(
            parsed.start.offset(&rest),
            undertest.len() - rest.fragment().len()
        );
        assert_eq!(*rest.fragment(), "and live to tell the tale");
    }

    #[test]
    fn test_parse_link() {
        let undertest = "[Foo](bar) and more";
        let expected = TokenType::Link(Link {
            text: "Foo",
            path: "bar",
        });
        let (rest, _parsed) = assert_parse_token_with_inner(undertest, parse_link, expected);
        assert_eq!(*rest.fragment(), " and more");
    }

    #[test]
    fn test_parse_block() {
        let undertest = "```\nfoo\nbar\n``` and more";
        let expected = TokenType::Block("\nfoo\nbar\n".into());
        let (rest, _parsed) = assert_parse_token_with_inner(undertest, parse_block, expected);
        assert_eq!(*rest.fragment(), " and more");
    }

    #[test]
    fn test_parse_inline() {
        let undertest = "`foo` and more";
        let expected = TokenType::Inline("foo".into());
        let (rest, _parsed) = assert_parse_token_with_inner(undertest, parse_inline, expected);
        assert_eq!(*rest.fragment(), " and more");
    }

    #[test]
    fn test_parse_keyword() {
        let undertest = "To \n";
        let expected = TokenType::Keyword(Keyword::To);
        let (rest, _parsed) = assert_parse_token_with_inner(undertest, parse_keyword, expected);
        assert_eq!(*rest.fragment(), "\n");
    }

    #[test]
    fn test_parse_newline() {
        let undertest = "\n";
        let expected = TokenType::Newline;
        let (rest, _parsed) = assert_parse_token_with_inner(undertest, parse_newline, expected);
        assert_eq!(*rest.fragment(), "");
    }

    pub(crate) fn assert_parse_token_with_inner<'a>(
        src: &'a str,
        parser: impl Fn(Span) -> IResult<Span, Token>,
        expected_inner: TokenType,
    ) -> (Span<'a>, Token<'a>) {
        let undertest = LocatedSpan::new(src);
        let (rest, parsed) = parser(undertest).unwrap();
        let _f: Result<
            (LocatedSpan<&str>, Token<'_>),
            nom::Err<nom::error::Error<LocatedSpan<&str>>>,
        > = parser(undertest);
        assert_eq!(parsed.inner, expected_inner);
        (rest, parsed)
    }
}
