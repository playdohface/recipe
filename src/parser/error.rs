use super::{Span, Token, TokenType};

#[derive(Debug)]
pub enum ParseError<'a> {
    Generic(&'a str),
    InvalidToken(Token<'a>),
    InvalidCodeBlock(&'a str),

    // Errors converting to a value
    InvalidValue(&'a str),
    InvalidLiteral((&'a str, json5::Error)),
    InvalidTokenTypeForValue(TokenType<'a>),

    InvalidSelection(nom::Err<nom::error::Error<&'a str>>),

    NomError(nom::error::VerboseError<Span<'a>>),
}
