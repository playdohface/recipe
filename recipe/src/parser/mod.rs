use std::collections::HashSet;

use nom::combinator::opt;
use nom::multi::many0;
use nom::{
    branch::alt,
    bytes::complete::{tag, take_while, take_while1},
    character::complete::{digit1, line_ending, space0, space1},
    combinator::{fail, verify},
    error::VerboseError,
    sequence::{delimited, preceded},
    IResult,
};
use nom_locate::LocatedSpan;
use serde::{Deserialize, Serialize};

use error::ParseError;
use tokenizer::Tokenizer;

use crate::context::Command;

mod error;
pub mod tokenizer;

pub type Span<'a> = LocatedSpan<&'a str>;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Heading<'a> {
    pub level: usize,
    pub text: &'a str,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct CodeBlock<'a> {
    pub executor: Option<&'a str>,
    pub name: Option<&'a str>,
    pub type_hint: Option<&'a str>,
    pub annotations: Vec<&'a str>,
    pub code: &'a str,
}

impl<'a> TryFrom<Token<'a>> for CodeBlock<'a> {
    type Error = ParseError<'a>;

    fn try_from(value: Token<'a>) -> Result<Self, Self::Error> {
        match value.inner {
            TokenType::Block(src) => parse_code_block(src)
                .map(|(_, res)| res)
                .map_err(|_| ParseError::InvalidCodeBlock(src)),
            _ => Err(ParseError::InvalidToken(value)),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Link<'a> {
    pub text: &'a str,
    pub path: &'a str,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum Selection<'a> {
    Index(usize),
    Key(&'a str),
}
impl<'a> Selection<'a> {
    fn as_string(&self) -> String {
        match self {
            Selection::Index(i) => i.to_string(),
            Selection::Key(k) => k.to_string(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct SelectionPath<'a>(#[serde(borrow)] pub Vec<Selection<'a>>);
impl<'a> SelectionPath<'a> {
    /// the selection as a JSON-Pointer per [RFC6901](https://datatracker.ietf.org/doc/html/rfc6901)
    pub fn json_pointer(&self) -> String {
        let mut pointer = String::new();
        for s in self.0.iter() {
            // both Tilde ~ and Slash / are disallowed in names so we can skip escaping
            pointer = format!("{pointer}/{}", s.as_string())
        }
        pointer
    }

    /// The last element
    pub fn last(&self) -> Option<&Selection> {
        self.0.last()
    }
    /// The selection up until the second-to-last element
    pub fn parent(&self) -> Option<SelectionPath> {
        if self.0.len() > 1 {
            let mut parent = self.0.clone();
            parent.pop();
            Some(SelectionPath(parent))
        } else {
            None
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum Val<'a> {
    ContextRef(#[serde(borrow)] SelectionPath<'a>),
    Literal(serde_json::Value),
}
impl<'a> TryFrom<&TokenType<'a>> for Val<'a> {
    type Error = ParseError<'a>;

    fn try_from(value: &TokenType<'a>) -> Result<Self, ParseError<'a>> {
        match *value {
            TokenType::Inline(src) => {
                if let Ok((_, val)) = parse_selection_path(src) {
                    Ok(Val::ContextRef(val))
                } else if let Ok(val) = json5::from_str(src) {
                    Ok(Val::Literal(val))
                } else {
                    Err(ParseError::InvalidValue(src))
                }
            }
            TokenType::Block(src) => match json5::from_str(src) {
                Ok(val) => Ok(Val::Literal(val)),
                Err(e) => Err(ParseError::InvalidLiteral((src, e))),
            },
            _ => Err(ParseError::InvalidTokenTypeForValue(value.clone())),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct SetDirective<'a> {
    #[serde(borrow)]
    pub variable: SelectionPath<'a>,
    pub value: Val<'a>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum Keyword {
    To,
    Set,
    Execute,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct Token<'a> {
    start: usize,
    end: usize,
    inner: TokenType<'a>,
}
impl<'a> Token<'a> {
    pub fn inner(&self) -> &TokenType<'a> {
        &self.inner
    }
    pub fn is_inline(&self) -> bool {
        matches!(self.inner, TokenType::Inline(_))
    }
    pub fn is_newline(&self) -> bool {
        self.inner == TokenType::Newline
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub enum TokenType<'a> {
    Heading(Heading<'a>),
    Block(&'a str),
    Inline(&'a str),
    Link(Link<'a>),
    Keyword(Keyword),
    Newline,
}

pub struct Tokens<'a> {
    tokens: Vec<Tokenizer<'a>>,
    loaded_paths: HashSet<String>,
}
impl<'a> Tokens<'a> {
    pub fn new() -> Self {
        Self {
            tokens: Vec::new(),
            loaded_paths: HashSet::new(),
        }
    }
    pub fn load(&mut self, src: String, file_path: &'a str, heading_slug: Option<&'a str>) {
        if self.loaded_paths.insert(format!(
            "{}#{}",
            file_path,
            heading_slug.unwrap_or_default()
        )) {
            //TODO: we may want to tie the lifetime of the loaded files to the lifetime of the Tokens struct or Tokenizer instead of just .leak()ing
            //TODO: we are also prepending a newline here to avoid missing headings at the start of the file, but this messes with our offsets in Token
            let mut tokenizer = Tokenizer::from_str(format!("\n{src}").leak(), file_path);
            if let Some(heading_slug) = heading_slug {
                tokenizer.scope_to_heading(heading_slug);
            }
            self.tokens.push(tokenizer);
        }
    }
    pub fn current_path(&self) -> Option<&'a str> {
        self.tokens.last().map(|t| t.origin_path)
    }
}
impl<'a> Default for Tokens<'a> {
    fn default() -> Self {
        Self::new()
    }
}
impl<'a> Iterator for Tokens<'a> {
    type Item = Token<'a>;

    fn next(&mut self) -> Option<Self::Item> {
        while let Some(tokenizer) = self.tokens.last_mut() {
            if let Some(token) = tokenizer.next() {
                return Some(token);
            } else {
                self.tokens.pop();
            }
        }
        None
    }
}

/// Parse a series of Commands from Tokens, i.e. the body of a To-Directive
pub fn parse_to_directive_inner<'a>(tokens: &mut Tokens<'a>) -> Vec<Command<'a>> {
    let mut commands = Vec::new();
    while let Some(token) = tokens.next() {
        match token.inner {
            TokenType::Block(block) => {
                if let Ok((_, codeblock)) = parse_code_block(block) {
                    commands.push(Command::CodeBlock(codeblock));
                } else {
                    todo!("Invalid Code Block");
                }
            }
            TokenType::Keyword(Keyword::Set) => {
                if let Ok(set_directive) = parse_set_directive_inner(tokens) {
                    commands.push(Command::SetDirective(set_directive));
                } else {
                    todo!("Invalid Set Directive");
                }
            }
            TokenType::Heading(_) => break,
            _ => continue, //TODO warnings/errors
        }
    }
    commands
}

fn parse_set_directive_inner<'a>(
    tokens: &mut Tokens<'a>,
) -> Result<SetDirective<'a>, ParseError<'a>> {
    if let (
        Some(Token {
            inner: TokenType::Inline(variable),
            ..
        }),
        Some(Token {
            inner: TokenType::Inline(value),
            ..
        }),
        Some(Token {
            inner: TokenType::Newline,
            ..
        }),
    ) = (tokens.next(), tokens.next(), tokens.next())
    {
        let (_, variable) = parse_selection_path(variable).map_err(ParseError::InvalidSelection)?;
        if let Ok((_, value)) = parse_selection_path(value) {
            Ok(SetDirective {
                variable,
                value: Val::ContextRef(value),
            })
        } else if let Ok(literal) = json5::from_str(value) {
            Ok(SetDirective {
                variable,
                value: Val::Literal(literal),
            })
        } else {
            Err(ParseError::InvalidValue(value))
        }
    } else {
        Err(ParseError::Generic("Wrong Tokentypes for Set-directive"))
    }
}
fn is_allowed_in_ident(c: char) -> bool {
    !c.is_whitespace() && c != '`'
}

fn is_allowed_in_name(c: char) -> bool {
    is_allowed_in_ident(c) && !['(', ')', '.', '[', ']', '{', '}', '\'', '"', '/', '~'].contains(&c)
}

fn parse_selection_path(inp: &str) -> IResult<&str, SelectionPath> {
    let (mut inp, first) = valid_name(inp)?;
    let mut sel_path = vec![Selection::Key(first)];
    while let Ok((rest, sel)) = alt((preceded(tag("."), parse_selection), parse_selection))(inp) {
        inp = rest;
        sel_path.push(sel);
    }
    Ok((inp, SelectionPath(sel_path)))
}

fn parse_selection(inp: &str) -> IResult<&str, Selection> {
    let mut array_index = delimited(tag("["), digit1::<_, VerboseError<&str>>, tag("]"));
    if let Ok((rest, idx)) = array_index(inp) {
        if let Ok(index) = idx.parse::<usize>() {
            return Ok((rest, Selection::Index(index)));
        } else {
            // arrays longer than u64::MAX will never be a thing in Recipe
            return fail(inp);
        }
    } else if let Ok((rest, name)) = valid_name(inp) {
        return Ok((rest, Selection::Key(name)));
    }
    fail(inp)
}

/// Parse the inner portion of a code-block
/// always consumes the entire input
/// The first word, not preceded by whitespace, is the executor
/// the next word, or first word preceded by whitespace is the name
/// the next word, if it is in parenthesis, is the type-hint
/// all following words until newline are annotations
/// everything after (but not including) the first newline is the code
fn parse_code_block(inp: &str) -> IResult<&str, CodeBlock> {
    let (inp, executor) = opt(valid_name)(inp)?;
    let (inp, name) = opt(preceded(space1, valid_name))(inp)?;
    let (inp, type_hint) = opt(preceded(space0, delimited(tag("("), valid_name, tag(")"))))(inp)?;
    let (inp, annotations) = many0(preceded(space1, valid_name))(inp)?;

    let (code, _) = preceded(space0, line_ending)(inp)?;
    let codeblock = CodeBlock {
        executor,
        name,
        type_hint,
        annotations,
        code,
    };
    let (rest, _) = take_while(|_| true)(code)?;
    Ok((rest, codeblock))
}

fn is_allowed_at_name_start(c: char) -> bool {
    is_allowed_in_name(c) && !['-'].contains(&c)
}

fn is_valid_name(name: &str) -> bool {
    name.starts_with(is_allowed_at_name_start)
}

fn valid_name(inp: &str) -> IResult<&str, &str> {
    verify(take_while1(is_allowed_in_name), is_valid_name)(inp)
}

#[cfg(test)]
mod tests {
    use super::*;

    impl<'a> TokenType<'a> {
        fn mock_token(self) -> Token<'a> {
            Token {
                start: 0,
                end: 0,
                inner: self,
            }
        }
    }

    impl<'a> CodeBlock<'a> {
        fn eq_by_value(&self, other: &CodeBlock) -> bool {
            if self.annotations.len() != other.annotations.len() {
                return false;
            }
            for i in 0..self.annotations.len() {
                if !(self.annotations[i] == other.annotations[i]) {
                    return false;
                }
            }
            self.code == other.code
                && self.name == other.name
                && self.executor == other.executor
                && self.type_hint == other.type_hint
        }
    }

    #[test]
    fn test_parse_code_block() {
        let undertest = r#"sh foo_bar (object) annot1 annot2
echo "I am a {{ banana }}"
"#;
        let expected = CodeBlock {
            executor: Some("sh".into()),
            name: Some("foo_bar".into()),
            type_hint: Some("object".into()),
            annotations: vec!["annot1".into(), "annot2".into()],
            code: "echo \"I am a {{ banana }}\"\n".into(),
        };

        let from_token: CodeBlock = TokenType::Block(undertest.into())
            .mock_token()
            .try_into()
            .unwrap();

        let (_, parsed) = parse_code_block(undertest.into()).unwrap();
        assert!(parsed.eq_by_value(&expected));
        assert!(from_token.eq_by_value(&expected));

        let undertest = r#"sh
echo "I am a {{ banana }}"
"#;
        let expected = CodeBlock {
            executor: Some("sh".into()),
            name: None,
            type_hint: None,
            annotations: vec![],
            code: "echo \"I am a {{ banana }}\"\n".into(),
        };

        let from_token: CodeBlock = TokenType::Block(undertest.into())
            .mock_token()
            .try_into()
            .unwrap();

        let (_, parsed) = parse_code_block(undertest.into()).unwrap();
        assert!(parsed.eq_by_value(&expected));
        assert!(from_token.eq_by_value(&expected));
    }

    #[test]
    fn test_parse_selection_path() {
        let src = "foo[2].bar and something else";
        let (rest, parsed) = parse_selection_path(src.into()).unwrap();
        assert_eq!(
            parsed,
            SelectionPath(vec![
                Selection::Key("foo"),
                Selection::Index(2),
                Selection::Key("bar")
            ])
        );
        assert_eq!(rest, " and something else");
    }

    #[test]
    fn test_parse_selection() {
        let src = "[0123].foo"; //if str.parse() accepts leading zeros, so do we
        let (rest, parsed) = parse_selection(src.into()).unwrap();
        assert_eq!(parsed, Selection::Index(123));
        assert_eq!(rest, ".foo");
        let src = "A_Valid_Key123[0]";
        let (rest, parsed) = parse_selection(src.into()).unwrap();
        assert_eq!(parsed, Selection::Key("A_Valid_Key123"));
        assert_eq!(rest, "[0]");
        assert!(parse_selection(".foo".into()).is_err());
        assert!(parse_selection("[3.14]".into()).is_err());
    }

    // #[test]
    // fn test_parse_to_directive() {
    //     let undertest =
    //         "To really `bake` a `cake` with `--flour`, `ignoreme` `--butter` and `--milk`: \r\nand more";
    //     let expected = TokenType::ToDirective(ToDirective {
    //         callpath: vec!["bake", "cake"],
    //         args: vec!["flour", "butter", "milk"],
    //     });
    //     let (rest, _) = assert_parse_token_with_inner(undertest, parse_to_directive, expected);
    //     assert_eq!(*rest.fragment(), "and more");
    //     assert!(parse_to_directive("To do nothing is no fun...\n".into()).is_err());
    //     assert!(parse_to_directive(
    //         "To have a `valid` to-directive without trailing newline is meaningless".into()
    //     )
    //     .is_err());
    //     let undertest = "To `have` `cake` without args is okay\n";
    //     let expected = TokenType::ToDirective(ToDirective {
    //         callpath: vec!["have", "cake"],
    //         args: vec![],
    //     });
    //     assert_parse_token_with_inner(undertest, parse_to_directive, expected);
    // }
}
