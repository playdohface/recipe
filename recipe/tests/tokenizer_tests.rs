use std::path::Path;

use insta::assert_yaml_snapshot;

use recipe::parser::{Heading, Keyword, Link, TokenType};
use recipe::parser::tokenizer::Tokenizer;

#[test]
fn test_scope_to_heading() {
    let src = r#"
# Heading 1
`inside heading 1`
## Heading 2 please ScopeMe-to_ThisHeading
`inside heading 2`
### Heading 3
`inside heading 3`
## Another heading
`inside another heading`
"#;
    let mut tokenizer = Tokenizer::from_str(src, Path::new("Mock.md"));
    tokenizer.scope_to_heading("heading-2-please-scope-me-to-this-heading");
    let tokens: Vec<_> = tokenizer.collect();
    assert_yaml_snapshot!(tokens);
}

#[test]
fn test_tokenizer() {
    let src = r#"
#Heading 1
Some text
```rust
### Not a heading
fn main() {
    println!("Hello, World!");
}
```
Some more text
[Link](https://example.com)
Set `foo` to `bar`
"#;
    let tokenizer = Tokenizer::from_str(src, Path::new("Mock.md"));
    let tokens: Vec<_> = tokenizer.collect();
    assert_yaml_snapshot!(tokens);
}
