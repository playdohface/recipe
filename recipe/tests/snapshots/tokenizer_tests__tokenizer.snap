---
source: tests/tokenizer_tests.rs
expression: tokens
snapshot_kind: text
---
- start: 0
  end: 12
  inner:
    Heading:
      level: 1
      text: Heading 1
- start: 21
  end: 22
  inner: Newline
- start: 25
  end: 96
  inner:
    Block: "rust\n### Not a heading\nfn main() {\n    println!(\"Hello, World!\");\n}\n"
- start: 96
  end: 97
  inner: Newline
- start: 111
  end: 112
  inner: Newline
- start: 112
  end: 139
  inner:
    Link:
      text: Link
      path: "https://example.com"
- start: 139
  end: 140
  inner: Newline
- start: 140
  end: 144
  inner:
    Keyword: Set
- start: 145
  end: 149
  inner:
    Inline: foo
- start: 154
  end: 158
  inner:
    Inline: bar
- start: 158
  end: 159
  inner: Newline
